File: coding/ocaml/code/zigzag.ml
let getLongestZigZag (inputs: int list) =
    let rec zigZag (inputs': int list) marker' count' isPeak' =
        match inputs' with
            | [] -> count' (* Done recursion, emit value *)
            | head::tail -> (let incr = marker' < head <> isPeak' in
              zigZag tail head (if incr && head <> marker' then count' + 1 else count') (if incr then not isPeak' else isPeak'))
    in
    match inputs with
    | a :: b :: tail -> zigZag ([b] @ tail) a 1 (a > b)
    | _ -> List.length inputs

File: coding/ocaml/code/zigzag.java
public static int getLongestZigZag(int[] inputs) {
    //Implement base cases
    if (inputs.length <= 2) return inputs.length;
    int marker = inputs[0];
    boolean isPeak = inputs[0] > inputs[1]; //set peak to accept first inputs
    int count = 1; //accept first number
    for (int i : inputs) {
        if (marker == i) continue;
        if (marker < i != isPeak) {
            count++;
            isPeak = !isPeak;
        }
        marker = i;
    }
    System.out.println(String.format("Input: %s\nCount: %d\n", Arrays.toString(inputs), count));
    return count;
}